internal/controller/iprange_controller.go

```diff
@@ -0,0 +1,306 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"maps"
+	"strconv"
+	"strings"
+
+	netboxv1 "github.com/netbox-community/netbox-operator/api/v1"
+	"github.com/netbox-community/netbox-operator/pkg/config"
+	"github.com/netbox-community/netbox-operator/pkg/netbox/api"
+	"github.com/netbox-community/netbox-operator/pkg/netbox/models"
+	"github.com/swisscom/leaselocker"
+	corev1 "k8s.io/api/core/v1"
+	apismeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const IpRangeFinalizerName = "iprange.netbox.dev/finalizer"
+const LastIpRangeMetadataAnnotationName = "iprange.netbox.dev/last-ip-range-metadata"
+
+// IpRangeReconciler reconciles a IpRange object
+type IpRangeReconciler struct {
+	client.Client
+	Scheme            *runtime.Scheme
+	NetboxClient      *api.NetboxClient
+	Recorder          record.EventRecorder
+	OperatorNamespace string
+	RestConfig        *rest.Config
+}
+
+//+kubebuilder:rbac:groups=netbox.dev,resources=ipranges,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=netbox.dev,resources=ipranges/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=netbox.dev,resources=ipranges/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *IpRangeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	logger.Info("reconcile loop started")
+
+	o := &netboxv1.IpRange{}
+	err := r.Client.Get(ctx, req.NamespacedName, o)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// if being deleted
+	if !o.ObjectMeta.DeletionTimestamp.IsZero() {
+		if o.Spec.PreserveInNetbox {
+			// if there's a finalizer remove it and return
+			// this can be the case if a CR used to have spec.preserveInNetbox set to false
+			return ctrl.Result{}, r.removeFinalizer(ctx, o)
+		}
+
+		err := r.NetboxClient.DeleteIpRange(o.Status.IpRangeId)
+		if err != nil {
+			err = r.logErrorSetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeReadyFalseDeletionFailed,
+				corev1.EventTypeWarning, "", err)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			return ctrl.Result{Requeue: true}, nil
```

Comment:
The logger which is set up in cmd/main.go will log the stack trace for every error returned by a reconcile function. Because we don't need the stack trace for netbox api errors we log the error in the reconcile loop directly and return nil.

Reply:
Closing as discussed